homeassistant/components/adguard/switch.py

```diff
@@ -0,0 +1,237 @@
+"""Support for AdGuard Home switches."""
+from datetime import timedelta
+import logging
+
+from homeassistant.components.adguard import AdGuardHomeDeviceEntity
+from homeassistant.components.adguard.const import (
+    DATA_ADGUARD_CLIENT, DATA_ADGUARD_VERION, DOMAIN)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.entity import ToggleEntity
+from homeassistant.helpers.typing import HomeAssistantType
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=10)
+PARALLEL_UPDATES = 1
+
+
+async def async_setup_entry(
+        hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up AdGuard Home switch based on a config entry."""
+    from adguardhome import AdGuardHomeConnectionError
+
+    adguard = hass.data[DOMAIN][DATA_ADGUARD_CLIENT]
+
+    try:
+        version = await adguard.version()
+    except AdGuardHomeConnectionError as exception:
+        raise PlatformNotReady from exception
+
+    hass.data[DOMAIN][DATA_ADGUARD_VERION] = version
+
+    switches = [
+        AdGuardHomeProtectionSwitch(adguard),
+        AdGuardHomeFilteringSwitch(adguard),
+        AdGuardHomeParentalSwitch(adguard),
+        AdGuardHomeSafeBrowsingSwitch(adguard),
+        AdGuardHomeSafeSearchSwitch(adguard),
+        AdGuardHomeQueryLogSwitch(adguard),
+    ]
+    async_add_entities(switches, True)
+
+
+class AdGuardHomeSwitch(ToggleEntity, AdGuardHomeDeviceEntity):
+    """Defines a AdGuard Home switch."""
+
+    def __init__(self, adguard, name: str, icon: str, key: str):
+        """Initialize AdGuard Home switch."""
+        self._state = False
+        self._key = key
+        super().__init__(adguard, name, icon)
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return '_'.join(
+            [
+                DOMAIN,
+                self.adguard.host,
+                str(self.adguard.port),
+                'switch',
+                self._key,
+            ]
+        )
+
+    @property
+    def is_on(self) -> bool:
+        """Return the state of the switch."""
+        return self._state
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn off the switch."""
+        from adguardhome import AdGuardHomeError
+
+        try:
+            await self._adguard_turn_off()
+        except AdGuardHomeError:
+            _LOGGER.error(
+                "An error occurred while turning off AdGuard Home switch."
+            )
+            self._available = False
+
+    async def _adguard_turn_off(self) -> None:
+        """Turn off the switch."""
+        raise NotImplementedError()
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn on the switch."""
+        from adguardhome import AdGuardHomeError
+
+        try:
```

Comment:
FYI, this could have also been achieved via a decorator.

Reply:
Aah true! Thanks, will improve on that.

Reply:
Took a second look. It would save me 3 lines of code in 2 files (so 6).

The price to pay is another level of "complexity" / code jump, which IMHO, in this case (saving is low) just harms the readability of the code.

While I do appreciate the suggestion, I would like to pass on this.